### PR TITLE
DPM-53: Updated DOSS to support authentication using OpenID Connect.

### PR DESCRIPTION
**Pull request by sjacob**
This change adds support for authenticating clients using OpenID Connect tokens.  It is implemented on top of SASL so potentially other authentication scheme could be bolted on in future.
I also updated the pom file so that a remote client specific jar is built along with the full doss jar file.
@ninhnguyen @ato @brycollins @scoen @terenceingram

### CODE REVIEW

**Comment by nlabuilder** [comment]
SonarQube analysis reported 23 issues
* CRITICAL 5 critical
* MAJOR 7 major
* MINOR 11 minor

#### Top 10 issues

1. CRITICAL [DOSS.java#L26](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L26): Make this "public static OIDCSERVER" field final [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1444)
1. CRITICAL [DOSS.java#L28](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L28): Make this "public static OIDCCLIENT" field final [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1444)
1. CRITICAL [DOSS.java#L30](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L30): Remove this hard-coded password. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2068)
1. CRITICAL [DOSS.java#L30](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L30): Make this "public static OIDCPASSWORD" field final [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1444)
1. CRITICAL [AuthenticatingTransportFactory.java#L43](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/security/AuthenticatingTransportFactory.java#L43): Define and throw a dedicated exception instead of using a generic one. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS00112)
1. MAJOR [DOSS.java#L26](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L26): Make OIDCSERVER a static final constant or non-public and provide accessors if needed. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)
1. MAJOR [DOSS.java#L28](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L28): Make OIDCCLIENT a static final constant or non-public and provide accessors if needed. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)
1. MAJOR [DOSS.java#L30](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/DOSS.java#L30): Make OIDCPASSWORD a static final constant or non-public and provide accessors if needed. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)
1. MAJOR [RemoteBlobStore.java#L92](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/net/RemoteBlobStore.java#L92): Make this anonymous inner class a lambda (sonar.java.source not set. Assuming 8 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [RemoteBlobStore.java#L107](https://github.com/nla/doss/blob/34083de3290e7f8fd8ba1e469e1805f4c525cbdd/src/doss/net/RemoteBlobStore.java#L107): Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2293)

**Comment by nlabuilder** [comment]
Build finished.

**Comment by nlabuilder** [comment]
SonarQube analysis reported 9 issues

1. CRITICAL [DOSS.java#L30](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/DOSS.java#L30): Remove this hard-coded password. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2068)
1. MAJOR [RemoteBlobStore.java#L92](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/net/RemoteBlobStore.java#L92): Make this anonymous inner class a lambda (sonar.java.source not set. Assuming 8 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [AuthenticatingTransportFactory.java#L28](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/security/AuthenticatingTransportFactory.java#L28): Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2293)
1. MAJOR [SaslOauthBearerCallbackHandler.java#L81](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/security/SaslOauthBearerCallbackHandler.java#L81): Extract the assignment out of this expression. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)
1. MINOR [DOSS.java#L26](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/DOSS.java#L26): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [DOSS.java#L28](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/DOSS.java#L28): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [DOSS.java#L30](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/DOSS.java#L30): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [AuthenticatingTransport.java#L35](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/security/AuthenticatingTransport.java#L35): Document this public constructor. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AUndocumentedApi)
1. MINOR [AuthenticatingTransportFactory.java#L30](https://github.com/nla/doss/blob/761c8a9be9965071b26a1e9e7bbf3087edeb636e/src/doss/security/AuthenticatingTransportFactory.java#L30): Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS00115)

**Comment by nlabuilder** [comment]
Build finished.

**Comment by nlabuilder** [comment]
Build finished.

File: pom.xml [hunk 1]
```diff
@@ -1,196 +1,204 @@
...
-    </dependency>
-    <dependency>
-        <groupId>ch.qos.logback</groupId>
-        <artifactId>logback-classic</artifactId>
-        <version>1.1.7</version>
-        <scope>test</scope>
-    </dependency>
-  </dependencies>
+	<dependencyManagement>
+		<dependencies>
```
Comment (shuangzhou) [inline pom.xml]: 👍

**Comment by nlabuilder** [comment]
SonarQube analysis reported 9 issues

1. CRITICAL [DOSS.java#L30](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/DOSS.java#L30): Remove this hard-coded password. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2068)
1. MAJOR [RemoteBlobStore.java#L92](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/net/RemoteBlobStore.java#L92): Make this anonymous inner class a lambda (sonar.java.source not set. Assuming 8 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [AuthenticatingTransportFactory.java#L28](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/security/AuthenticatingTransportFactory.java#L28): Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2293)
1. MAJOR [SaslOauthBearerCallbackHandler.java#L81](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/security/SaslOauthBearerCallbackHandler.java#L81): Extract the assignment out of this expression. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)
1. MINOR [DOSS.java#L26](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/DOSS.java#L26): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [DOSS.java#L28](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/DOSS.java#L28): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [DOSS.java#L30](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/DOSS.java#L30): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [AuthenticatingTransport.java#L35](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/security/AuthenticatingTransport.java#L35): Document this public constructor. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AUndocumentedApi)
1. MINOR [AuthenticatingTransportFactory.java#L30](https://github.com/nla/doss/blob/f3ce48a1d1860b9f2499cb6d4fb803f3dc3c7dd5/src/doss/security/AuthenticatingTransportFactory.java#L30): Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS00115)

**Comment by nlabuilder** [comment]
Build finished.

File: pom.xml [hunk 2]
```diff
@@ -1,196 +1,163 @@
...
+    <modelVersion>4.0.0</modelVersion>
 
-  <groupId>au.gov.nla</groupId>
-  <artifactId>doss</artifactId>
-  <version>1.1.2-SNAPSHOT</version>
-  <packaging>jar</packaging>
+    <parent>
+        <groupId>au.gov.nla</groupId>
+        <artifactId>doss-project</artifactId>
+        <version>1.1.2-SNAPSHOT</version>
```
Comment (brycollins) [inline pom.xml]: I think this qualifies as a new version, maybe 1.2

**Comment by nlabuilder** [comment]
SonarQube analysis reported 9 issues

1. CRITICAL [DOSS.java#L30](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/DOSS.java#L30): Remove this hard-coded password. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2068)
1. MAJOR [RemoteBlobStore.java#L92](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/net/RemoteBlobStore.java#L92): Make this anonymous inner class a lambda (sonar.java.source not set. Assuming 8 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1604)
1. MAJOR [AuthenticatingTransportFactory.java#L28](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/security/AuthenticatingTransportFactory.java#L28): Replace the type specification in this constructor call with the diamond operator ("<>"). (sonar.java.source not set. Assuming 7 or greater.) [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS2293)
1. MAJOR [SaslOauthBearerCallbackHandler.java#L81](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/security/SaslOauthBearerCallbackHandler.java#L81): Extract the assignment out of this expression. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)
1. MINOR [DOSS.java#L26](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/DOSS.java#L26): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [DOSS.java#L28](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/DOSS.java#L28): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [DOSS.java#L30](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/DOSS.java#L30): Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS1213)
1. MINOR [AuthenticatingTransport.java#L35](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/security/AuthenticatingTransport.java#L35): Document this public constructor. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AUndocumentedApi)
1. MINOR [AuthenticatingTransportFactory.java#L30](https://github.com/nla/doss/blob/0264e2e07b28d0cdf84895f8ed77bb313cb2903c/src/doss/security/AuthenticatingTransportFactory.java#L30): Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarqube.nla.gov.au/coding_rules#rule_key=squid%3AS00115)

**Comment by nlabuilder** [comment]
Build finished.

**Comment by ninhnguyen** [comment]
👍

**Pull request merged**